Array access projection

dario23 opened the pull request.

dario23 committed:
Handle ProjectionElem::Index in MirEncoder

dario23 committed:
Implement array access projection handling

dario23 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,7 +5496,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+
+                let array_type = base.get_type().clone();
+                let lookup_res: vir::Expr = self.get_auxiliary_local_var("lookup_ret", array_elem_ty).into();
+                let lookup_res_val_int = lookup_res.clone().field(vir::Field::new("val_int", vir::Type::Int));
```
this seems more manual than it needs to be, but i currently don't see how to do this better. it boils down to

 - the temporary variable seems to have to be type `Ref`, because the assignment to the final variable generated wants to copy from `.val_int`
 - that seems to mean that we need to pass `.val_int` to `lookup_pure`

fpoli replied:
Yes, as long as pure functions don't use snapshots for the parameters we have to add the `.val_int`. The field creation with the hardcoded string can be replaced with a call to `Encoder::encode_value_field`.

fpoli replied:
The temporary variable is of Viper type `Ref` because of `array_elem_ty`. By default we encode primitive types as heap-allocated values to simplify the encoding of creation of references. One of the future optimizations  (purification) should optimize cases like this.

dario23 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5477,6 +5540,17 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
 }
 
+// TODO: is there a better way? or even place to put this? type_encoder?
+fn rust_type_name_to_vir_type(rust_type_name: &str) -> EncodingResult<vir::Type> {
```
i'm really unhappy with this function, and still think i'd rather add the return type of lookup_pure to the `PlaceEncoding::ArrayAccess` variant, as when constructing that we still have access to the rust type, which allows simple generation of both, where now we have to do these name-based hacks..

dario23 replied:
so this would be what i'd prefer: https://github.com/dario23/prusti-dev/compare/array-access-projection...dario23:lookup-pure-ret?expand=1

fpoli replied:
I would also prefer to add one more field to avoid this function.

dario23 committed:
Pass lookup_pure_ret through in PlaceEncoding::ArrayAccess

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5467,7 +5496,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                let array_type = base.get_type().clone();
+                let lookup_res: vir::Expr = self.get_auxiliary_local_var("lookup_ret", array_elem_ty).into();
+                let lookup_res_val_int = lookup_res.clone().field(vir::Field::new("val_int", vir::Type::Int));
+                let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(*base)?;
```
I think there is a problem when you have multiple nested `ArrayAccess`, because the `get_auxiliary_local_var` call two lines above is designed to reuse generated variables but you always want a fresh one. Also, when you encode an expression you don't know for sure what's the last time that the expression will be used; it's possible that `postprocess_place_encoding` will be called again perfore that and (incorrectly) reuse the same temporary variable. You could implement something like a new `ProcedureEncoder::generate_fresh_local_var(..)` that uses an internal counter to generate new variable names.

dario23 committed:
Add val_field to PlaceEncoding::ArrayAccess

val_field is encoded from the rust type, which we no longer have later
on, so we create it in mir_encoder while we still can.

dario23 committed:
Use fresh local var in array access postprocessing

dario23 committed:
Thread errors from encode_loan_places through

This way we don't just panic, but return EncodingError::unsupported
properly

dario23 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5496,10 +5504,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                stmts.extend(self.encode_havoc_and_allocation(&lookup_res));
+                let lookup_pure_call = vir::Expr::func_app(
+                    lookup_pure,
+                    vec![encoded_base_expr, index.field(val_field)],
```
NOTE to self: just val_int here, index should always be `usize`

dario23 committed:
Array index is always of type `usize`

dario23 committed:
Clarify unsupported message

fpoli commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -5496,10 +5504,41 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
+                stmts.extend(self.encode_havoc_and_allocation(&lookup_res));
+                let lookup_pure_call = vir::Expr::func_app(
+                    lookup_pure,
+                    vec![encoded_base_expr, index.field(vir::Field::new("val_int", vir::Type::Int))],
```
In case you need to build thinks like `index.field(vir::Field::new("val_int", vir::Type::Int))` again, `Encoder::encode_value_expr(.., tcx.mk_ty(..))` might be slightly better. (No need to change this PR, snapshots are going to change this anyway.)

fpoli reviewed (approved):
LGTM
